krabcake/rs_hello/src/lib.rs

```diff
@@ -1,9 +1,51 @@
 #![feature(core_intrinsics, lang_items, c_size_t)]
+#![feature(c_unwind)]
 #![no_std]
 #![allow(unused_imports)]
 
-use core::ffi::{c_char,c_int,CStr,c_size_t,c_void};
+use core::ffi::{c_char, c_int, c_size_t, c_void, CStr};
 use core::panic::PanicInfo;
+use core::ptr;
+
+extern crate alloc;
+use core::alloc::{GlobalAlloc, Layout};
+
+// Can now import and use anything in alloc
+// use alloc::vec::Vec;
+
+extern "C" {
+    // From krabcake-vg/include/pub_tool_mallocfree.h
+    // Nb: the allocators *always succeed* -- they never return NULL (Valgrind
+    // will abort if they can't allocate the memory).
+    // The 'cc' is a string that identifies the allocation point.  It's used when
+    // --profile-heap=yes is specified.
+    // extern void* VG_(malloc)         ( const HChar* cc, SizeT nbytes );
+    // extern void  VG_(free)           ( void* p );
+    //
+    // Basic types are in krabcake-vg/VEX/pub/libvex_basictypes.h
+    // HChar = C char type
+    // Implementation lives in krabcake-vg/coregrind/m_mallocfree.c
+    fn vgPlain_malloc(cc: *const c_char, nbytes: c_size_t) -> *mut c_void;
+    fn vgPlain_free(p: *mut c_void);
+}
+
+struct ValgrindAllocator;
+
+#[global_allocator]
+static ALLOCATOR: ValgrindAllocator = ValgrindAllocator;
+
+unsafe impl Sync for ValgrindAllocator {}
+
+unsafe impl GlobalAlloc for ValgrindAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let cc: &[u8] = b"cc"; // TODO: Set a unique identifier here?
```

Comment:
bryan: Does this need to be null-terminated? I.e. `b"cc\0"` ? IIRC Rust won't do that for you and I cannot guess how vgPlain_malloc is going to figure out how long the string is otherwise.